### PR TITLE
daml deploy, now with party management

### PR DESCRIPTION
**Pull request by nickchapman-da**
WIP: `daml deploy` now with party management
- looking for feedback from @associahedron and Soren Bleikertz

DONE:
- new commands: `daml list-parties`, `daml allocate-party`
- `daml list-parties` reports the parties already allocated on the ledger
- `daml allocate-party` is idempotent, only allocating a new party when it does not already exist
- `daml deploy` is extended to allocate all parties in the config (idempotent) as well as upload the dar
- All commands take `--host` and `--port` flags
- All commands consult `daml.yaml` for `ledger-host` and `ledger-port` settings

TODO:
- navigator startup should take account of mapping shown by `daml list-parties`

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: release/sdk-config.yaml.tmpl [hunk 1]
```diff
@@ -19,10 +19,22 @@ commands:
...
+- name: list-parties
+  path: daml-helper/daml-helper
+  desc: "List parties known to the ledger"
+  args: ["list-parties"]
+
+- name: allocate-party
+  path: daml-helper/daml-helper
+  desc: "Allocate named party on the ledger"
+  args: ["allocate-party"]
+
```
Comment (associahedron) [inline release/sdk-config.yaml.tmpl]: I don't think these should both be top level commands. Either we have:

* `daml party` with subcommands `daml party list` and `daml party allocate`, or
* these are all subcommands of `daml deploy`, as is the original `daml deploy` (could be `daml deploy package`, `daml deploy party`, `daml deploy party --list`, or so).

I'm partial to the latter (i.e. having all of these be subcommands of `daml deploy`), since the thing all of these commands have in common is that they need to communicate with a ledger on a given host/port.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs [hunk 2]
```diff
@@ -81,6 +84,9 @@ required msg = fromMaybeM (throwIO $ DamlHelperError msg Nothing)
 requiredE :: Exception e => T.Text -> Either e t -> IO t
 requiredE msg = fromRightM (throwIO . DamlHelperError msg . Just . T.pack . displayException)
 
+defaulting :: a -> Either ConfigError (Maybe a) -> IO a
+defaulting a e = fmap (fromMaybe a) $ either throwIO return e
```
Comment (associahedron) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: Prefer `requiredE` which is defined above, results in better error messages.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs [hunk 3]
```diff
@@ -711,6 +753,26 @@ getProjectName = do
...
+    requiredE "Project must have parties listed" $
+        queryProjectConfigRequired ["parties"] projectConfig
+
+-- TODO: `daml sandbox` should also consult the config for the ledger-port
+-- Have daml-helper wrap the `sandbox` command
+getProjectLedgerPort :: IO Int
+getProjectLedgerPort = do
+    projectConfig <- getProjectConfig
+    -- TODO: remove default; insist ledger-port is in the config ?!
+    defaulting 6865 $ queryProjectConfig ["ledger-port"] projectConfig
```
Comment (associahedron) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: I think it would be better to use `queryProjectConfig ["deploy", "port"]` here, and `queryProjectConfig ["deploy", "host"]` below.

File: release/sdk-config.yaml.tmpl (hunk 1 above)
Comment (nickchapman-da) [reply release/sdk-config.yaml.tmpl]: I'm open to the idea of structured commands. Is there any prior art for this in `daml` ?

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs [hunk 4]
```diff
@@ -672,23 +680,57 @@ runStart sandboxPort (StartNavigator shouldStartNavigator) (OpenBrowser shouldOp
...
-runDeploy :: String -> SandboxPort -> IO ()
-runDeploy host sandboxPort = do
+data HostAndPortFlags = HostAndPortFlags { hostM :: Maybe String, portM :: Maybe Int }
+
+withHostAndPortDefaults :: HostAndPortFlags -> IO HostAndPort
+withHostAndPortDefaults HostAndPortFlags{hostM,portM} = do
+    defaultHost <- getProjectLedgerHost
+    defaultPort <- getProjectLedgerPort
+    let host = fromMaybe defaultHost hostM
+    let port = fromMaybe defaultPort portM
```
Comment (associahedron) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: In both of these cases, you only really want to get the default host and port if they weren't specified, right? I think `fromMaybeM getProjectLedgerHost hostM` and `fromMaybeM getProjectLedgerPort portM` do what we want here.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 2 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: There is something odd about using `requiredE` for non required config.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs [hunk 5]
```diff
@@ -48,6 +45,8 @@ commandParser =
          , command "init" (info (initCmd <**> helper) idm)
          , command "start" (info (startCmd <**> helper) idm)
          , command "deploy" (info (deployCmd <**> helper) idm)
+         , command "list-parties" (info (listPartiesCmd <**> helper) idm)
+         , command "allocate-party" (info (allocatePartyCmd <**> helper) idm)
```
Comment (associahedron) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs]: Like I said beloy, I think these would all be better organized as subcommands of "deploy".

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 3 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: Ok. Do we already make use of this kind hierarchical config anywhere else?

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs (hunk 5 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs]: Yup. Where are the existing examples?

File: release/sdk-config.yaml.tmpl (hunk 1 above)
Comment (associahedron) [reply release/sdk-config.yaml.tmpl]: `da` used to have a bunch of them, and `daml install` sort of simulates this (with `daml install latest` / `daml install project` / `daml install VERSION`) but doesn't really do it the proper way, which is by using `subparser` and `command` in optparse-applicative.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 2 above)
Comment (associahedron) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: It throws an error on the error part, not the config part ... just like `defaulting`.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 4 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: ok!

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 3 above)
Comment (associahedron) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: Not yet, but it's probably a bad thing that the project config file is so unstructured at the moment. E.g. build config really ought to be separate from project metadata.

Hierarchy is used in the SDK config to determine commands that are available, but it doesn't use `querySdkConfig` to get the individual command data (it couldn't, since it's a list).

Comment (associahedron) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: Here's the right way to do it:

Define a `DeployConfig` data structure. Implement `FromJSON` for it. Then use `queryProjectConfig ["deploy"]` to get it.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 2 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: I was having trouble seeing what config error could occur... given that the requirement for the field is optional. But I suppose the field might exist but be of the wrong type...

Which now makes more sense of the exiting code...
```
 mbScenario :: Maybe String <-
        requiredE "Failed to parse scenario" $
        queryProjectConfig ["scenario"] projectConfig
```

ok, I will change to be the same.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs (hunk 5 above)
Comment (associahedron) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs]: I can dig up an old example for you.

Comment (associahedron) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs]: Here, from the `da` code base:

https://github.com/digital-asset/daml/blob/1fcf2e4442f102c894af06164c5c0872db85e4b0/da-assistant/DA/Sdk/Cli/Command.hs#L107

You can see `templateActionParser` is defined further below as another `subparser ... subcommand` list. Note that `subcommand` is defined in that file as a convenience around `command ... (info ...)` but you don't really need it.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 2 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: done

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 4 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: done

**Comment by nickchapman-da** [comment]
@associahedron 
Do you have enough now in place to implement  the navigator fix?
I imagine you will call `DA.Daml.Helper.Ledger.listParties` directly.
Will you do that work in this PR, or a subsequent?

**Comment by associahedron** [comment]
> @associahedron
> Do you have enough now in place to implement the navigator fix?
> I imagine you will call `DA.Daml.Helper.Ledger.listParties` directly.
> Will you do that work in this PR, or a subsequent?

A subsequent PR. It's not really a navigator fix so much as a workaround. I think calling `daml deploy navigator` to open the navigator at the configured host & port is the right interface for this, and the workaround would only take place when the navigator is invoked through there (and need only be in place until navigator is updated to fetch the list of parties by itself).

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs [hunk 6]
```diff
@@ -672,23 +680,55 @@ runStart sandboxPort (StartNavigator shouldStartNavigator) (OpenBrowser shouldOp
...
-            exitFailure
+    Ledger.uploadDarFile hp bytes
+    putStrLn "Deploy succeeded."
+    exitSuccess
+
+
+allocatePartyIfRequired :: HostAndPort -> String -> IO ()
+allocatePartyIfRequired hp name = do
+    party <-
+        Ledger.lookupParty hp name >>= \case
```
Comment (associahedron) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: ah, ocaml...

I think the more idiomatic phrasing would be:

```Haskell
partyM <- Ledger
party <- flip fromMaybeM partyM $ do
    putStrLn $ "Allocating party for '" <> name <> "' at " <> show hp
    Ledger.allocateParty hp name
```

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs [hunk 7]
```diff
@@ -711,6 +751,28 @@ getProjectName = do
     requiredE "Project must have a name" $
         queryProjectConfigRequired ["name"] projectConfig
 
+getProjectParties :: IO [String]
+getProjectParties = do
+    projectConfig <- getProjectConfig
+    requiredE "Project must have parties listed" $
```
Comment (associahedron) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: The error message seems incorrect a bit off, because it might appear if the parties are there but are ill formed, and only the left side of the Either has enough information to disambiguate the issue. I think something like `"Failed to read list of parties from project config"` would be better.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs [hunk 8]
```diff
@@ -0,0 +1,51 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+module DA.Daml.Helper.Ledger (
```
Comment (associahedron) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs]: I like the separate module :-) 

I'd also be happy if all of the related logic for running the commands was in this module as well (the `runDeploy`, `runAllocateParty`, `runListParty`, etc). But happy to leave this refactoring for another PR. DamlHelper sure needs it at this point.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    xs <- listParties hp
+    let text = Text.pack name
+    let pred PartyDetails{displayName,party} = if text == displayName then Just party else Nothing
+    return $ List.firstJust pred xs
+
+allocateParty :: HostAndPort -> String -> IO Party
+allocateParty hp name = run hp $ do
+    let text = Text.pack name
+    let request = L.AllocatePartyRequest
+            { partyIdHint = "" -- text -- ???
```
Comment (soren-da) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs]: Yes we should set the party name as a hint. Sandbox can take it as is and in Canton we would get more readable party ids.

File: release/sdk-config.yaml.tmpl (hunk 1 above)
Comment (nickchapman-da) [reply release/sdk-config.yaml.tmpl]: I  think we can leave this restructuring to a later PR. It will be nice to get more opinions form the produce people.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 3 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: Again,  we can postpone this kind of restructuring to later.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs (hunk 9 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs]: Why do we even care that party names are readable?

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 6 above)
Comment (hurryabit) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: I wish we had `whenNothing :: Applicative f => Maybe a -> f a -> f a`. @ndmitchell would you accept that for the `extra` package?

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs [hunk 10]
```diff
@@ -672,23 +680,55 @@ runStart sandboxPort (StartNavigator shouldStartNavigator) (OpenBrowser shouldOp
...
+    putStrLn $ "Checking party allocation at " <> show hp
+    allocatePartyIfRequired hp name
+    exitSuccess
+
+runDeploy :: HostAndPortFlags -> IO ()
+runDeploy flags = do
+    hp <- withHostAndPortDefaults flags
+    putStrLn $ "Deploying to " <> show hp
+    parties <- getProjectParties
+    mapM_ (allocatePartyIfRequired hp) parties
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: Is there a way to do a batch allocation?

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs [hunk 11]
```diff
@@ -672,23 +680,55 @@ runStart sandboxPort (StartNavigator shouldStartNavigator) (OpenBrowser shouldOp
                   then withNavigator
                   else (\_ _ _ f -> f sandboxPh)
 
-runDeploy :: String -> SandboxPort -> IO ()
-runDeploy host sandboxPort = do
+data HostAndPortFlags = HostAndPortFlags { hostM :: Maybe String, portM :: Maybe Int }
+
+withHostAndPortDefaults :: HostAndPortFlags -> IO HostAndPort
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: To me a function called `withX` suggests that it wants a continuation as one of its arguments, most likely the last one. Renaming this to `getX` might be a good idea.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs [hunk 12]
```diff
@@ -672,23 +680,55 @@ runStart sandboxPort (StartNavigator shouldStartNavigator) (OpenBrowser shouldOp
...
+    host <- fromMaybeM getProjectLedgerHost hostM
+    port <- fromMaybeM getProjectLedgerPort portM
+    return HostAndPort {host,port}
+
+runListParties :: HostAndPortFlags -> IO ()
+runListParties flags = do
+    hp <- withHostAndPortDefaults flags
+    putStrLn $ "Listing parties at " <> show hp
+    xs <- Ledger.listParties hp
+    when (null xs) $ putStrLn "no parties are known"
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: I think
```suggestion
    if null xs then putStrLn "no parties are known" else mapM_ print xs
```
and removing the next line is clearer.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs (hunk 9 above)
Comment (soren-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs]: It makes debugging easier if the party ids are readable to some degree. I'd appreciate if we set the hint based on the name.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 6 above)
Comment (ndmitchell) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: concern is that I'd expect whenNothing to be when (isNothing x) ... and this doesn't have that type. flip fromMaybeM seems quite nice here, without inventing any new names.

Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: Happy to change to Fran's suggestion

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 7 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: This message was modelled of the exiting message in the code:
```
getProjectName :: IO String
getProjectName = do
    projectConfig <- getProjectConfig
    requiredE "Project must have a name" $
        queryProjectConfigRequired ["name"] projectConfig
```

Comment (associahedron) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: I see. That message has the same problem. Please change it to `"Failed to read project name from project config"` or so.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 10 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: `daml deploy` does batch allocation of everything in the `daml.yaml` config.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 6 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: depends what you idioms are :)
but I'll change it

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs (hunk 9 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs]: done.. next PR

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 10 above)
Comment (hurryabit) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: I meant batch allocation over the Ledger API. This looks like we're checking for each individual party if it already exists and allocate it when it doesn't. That's `O(n)` API calls for allocating `n` parties. I'm wondering if there's a way to do it in `O(1)` API calls? ("No" is a perfectly fine answer.)

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 7 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: will fix both messages in next PR

Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: @associahedron We could be more DRY here. The config lookup combinators should automate these messages.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 11 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: ok, will do

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 12 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: man, we are getting dangerously close to the bike shed here, but I will do it...

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs (hunk 10 above)
Comment (nickchapman-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: "No", the API does not support batch allocation.

Comment (hurryabit) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Run.hs]: Thanks.
